File: death/loop.py
"""Death detection and resurrection system.

``DeathLoop`` is stateless — both methods take a ``GlobalBlackboard`` and
either read from it (``check_death``) or mutate it (``resurrect``).

Death occurs when any stat reaches 0 (depleted) or 100 (overloaded).
Resurrection carries non-temporary tags forward as *karma* and resets all
other transient state so the player starts a new life with history intact.
"""

from __future__ import annotations

from pydantic import BaseModel

from game.state import GlobalBlackboard


class DeathInfo(BaseModel):
    """Snapshot of the game state at the moment of death."""

    cause_stat: str          # ID of the stat that crossed a boundary
    cause_value: int         # The boundary value (0 = depleted, 100 = overloaded)
    turn: int                # Turn number within the current week
    life_number: int         # Which life this was (1-based)
    tags_at_death: list[str] # Tags held when death occurred
    stats_at_death: dict[str, int]  # Stat values at the moment of death


class DeathLoop:
    """Stateless helper for death detection and resurrection."""

    @staticmethod
    def check_death(state: GlobalBlackboard) -> DeathInfo | None:
        """Return a ``DeathInfo`` if any stat is at 0 or 100, else None."""
        for stat_id, value in state.stats.items():
            if value <= 0:
                return DeathInfo(
                    cause_stat=stat_id,
                    cause_value=0,
                    turn=state.turn,
                    life_number=state.life_number,
                    tags_at_death=list(state.tags),
                    stats_at_death=dict(state.stats),
                )
            if value >= 100:
                return DeathInfo(
                    cause_stat=stat_id,
                    cause_value=100,
                    turn=state.turn,
                    life_number=state.life_number,
                    tags_at_death=list(state.tags),
                    stats_at_death=dict(state.stats),
                )
        return None

    @staticmethod
    def resurrect(state: GlobalBlackboard) -> list[str]:
        """Reset state for a new life and return the karma tags carried forward.

        Karma = non-temporary tags (anything not prefixed with ``_temp``), capped
        at 10.  Karma is added to ``state.karma`` for historical tracking and
        kept in ``state.tags`` so the new life starts with those modifiers.

        Stats are reset to 50 and the NPC appearance counters are zeroed so the
        Writer produces a fresh distribution of card characters.
        """
        important_tags = [t for t in state.tags if not t.startswith("_temp")]
        karma = important_tags[:10]

        state.previous_life_tags = karma.copy()
        state.karma.extend(karma)
        state.life_number += 1

        # Keep karma tags
        state.tags = set(karma)

        # Reset stats to 50
        for stat_id in state.stats:
            state.stats[stat_id] = 50

        # Reset NPC appearances
        for npc in state.npcs:
            npc.npc_appearance_count = 0

        # Skip time to next season
        from game.state import SEASONS_PER_YEAR
        state.season_index = (state.season_index + 1) % SEASONS_PER_YEAR
        if state.season_index == 0:
            state.year += 1
            
        state.day = 1
        state.turn = 0

        return karma

File: server/internal/death/loop.go
package death

// DeathInfo contains information about a death event
type DeathInfo struct {
	CauseStat string            `json:"cause_stat"`
	Turn      int               `json:"turn"`
	LifeNumber int              `json:"life_number"`
	Tags      map[string]bool   `json:"tags"`
	Stats     map[string]int    `json:"stats"`
}

// GameState is an interface for game state operations
type GameState interface {
	GetElapsedDays() int
	GetStats() map[string]int
	GetTags() map[string]bool
	GetNPCIDs() []string
	DisableNPC(id string)
	ClearEvents()
	SetIsAlive(alive bool)
	SetDeathCause(cause string)
	SetDeathTurn(turn int)
	SetSeason(season int)
	SetYear(year int)
	SetDay(day int)
	SetTags(tags map[string]bool)
	SetCurrentLife(life int)
}

// DeathLoop handles death detection and resurrection
type DeathLoop struct {
	state GameState
}

// NewDeathLoop creates a new death loop
func NewDeathLoop(state GameState) *DeathLoop {
	return &DeathLoop{state: state}
}

// CheckDeath detects when any stat hits 0 or 100
func (dl *DeathLoop) CheckDeath() (*DeathInfo, bool) {
	stats := dl.state.GetStats()
	for statID, value := range stats {
		if value <= 0 || value >= 100 {
			deathInfo := &DeathInfo{
				CauseStat:  statID,
				Turn:       dl.state.GetElapsedDays(),
				LifeNumber: 1, // Will be set by caller
				Tags:       make(map[string]bool),
				Stats:      make(map[string]int),
			}

			// Copy current state
			for k, v := range dl.state.GetTags() {
				deathInfo.Tags[k] = v
			}
			for k, v := range stats {
				deathInfo.Stats[k] = v
			}

			dl.state.SetIsAlive(false)
			dl.state.SetDeathCause(statID)
			dl.state.SetDeathTurn(dl.state.GetElapsedDays())

			return deathInfo, true
		}
	}

	return nil, false
}

// Resurrect resets world for new life
func (dl *DeathLoop) Resurrect(tempTags map[string]bool) {
	// Keep non-temp tags as "karma" (up to 10)
	karmaTags := make(map[string]bool)
	count := 0
	for tagID, active := range dl.state.GetTags() {
		if active && !tempTags[tagID] && count < 10 {
			karmaTags[tagID] = true
			count++
		}
	}

	// Reset stats to 50
	stats := dl.state.GetStats()
	for statID := range stats {
		stats[statID] = 50
	}

	// Clear NPC appearances
	for _, npcID := range dl.state.GetNPCIDs() {
		dl.state.DisableNPC(npcID)
	}

	// Clear events
	dl.state.ClearEvents()

	// Advance to next season
	season := 0 // Will be set by caller
	year := 0   // Will be set by caller
	season++
	if season > 3 {
		season = 0
		year++
	}

	dl.state.SetSeason(season)
	dl.state.SetYear(year)
	dl.state.SetDay(1)

	// Reset tags to karma only
	dl.state.SetTags(karmaTags)

	// Update life counter
	dl.state.SetCurrentLife(1) // Will be incremented by caller
	dl.state.SetIsAlive(true)
	dl.state.SetDeathCause("")
	dl.state.SetDeathTurn(0)
}
